File: binarysearch/Palindrome-Linked-List/Palindrome-Linked-List.cpp
// https://helloacm.com/two-algorithms-to-determine-a-palindrome-linked-list-using-a-stack/
// https://binarysearch.com/problems/Palindrome-Linked-List
// MEDIUM, STACK, LINKED LIST

/**
 * class LLNode {
 *     public:
 *         int val;
 *         LLNode *next;
 * };
 */
bool solve(LLNode* node) {
    if (!node) return true;
    stack<LLNode*> st;
    LLNode* head = node;
    while (head) {
        st.push(head);
        head = head->next;
    }
    while (!st.empty()) {
        auto p = st.top();
        st.pop();
        if (p->val != node->val) return false;
        node = node->next;
    }
    return true;
}

File: binarysearch/Palindrome-Linked-List/Palindrome-Linked-List.py
# https://helloacm.com/teaching-kids-programming-algorithms-to-check-if-a-linked-list-is-palindrome/
# https://binarysearch.com/problems/Palindrome-Linked-List
# MEDIUM, RECURSION

# class LLNode:
#     def __init__(self, val, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def solve(self, node):
        root = node
        def dfs(node):
            if node is None:
                return True
            nonlocal root
            ans = dfs(node.next)
            if root.val != node.val:
                return False
            root = root.next
            return ans
        return dfs(node)
